### DISCUSSION TITLE
Making completions more smarter

### DISCUSSION THREAD

**Discussion by gayanper** [eclipse-jdt/eclipse.jdt.core#780] [Ideas]
Hi All,

I have a small idea we could easily make JDT completions more smarter. To explain my idea I will take the following example

```
Stream.of(customers).filter(c -> c.isActive()).collect(|
```

Now at this place, the completion engine doesn't provide much help to the developer unless they type `Collectors.`. So what if we can tell the CompletionEngine to look at some class's accessible members that might match the target parameter ?

So in this case if we can tell look into `Collectors` class, that will result in some good suggestions. And in vscode this will be more refined by the microsoft intellicode ranking.

So to do that, I suggest we introduce a new settings similar to favorites to hold such classes as a simpler solution.

WDYT ?

**Comment by gayanper** [comment]
@mickaelistria @rgrunber @testforstephen @vogella @jdneo Please provide your ideas as well.

**Comment by jdneo** [comment]
@eskibear Is this the case where intelliCode can help?

**Reply by Eskibear** [reply]
No. Intellicode-Java currently helps ranking member methods when you type `SomeClass.|` or `someVariable.|`

**Comment by mickaelistria** [comment]
I believe it's an occurrence of https://bugs.eclipse.org/bugs/show_bug.cgi?id=567556 .

**Reply by gayanper** [reply]
Its not exactly chain completion’s, but yes we can use the chain completion implementation if we can some how provide classes like Collectors into it.

**Comment by gayanper** [comment]
I think we can move the main chain completion logic into jdt.core.manupalation module. Then reuse it in jdt.core. At the same time we can use it in jdt.ls when the completions are triggered CompletionTriggerKind.Invoked. Also we could try to find chain completions are returns them for CompletionTriggerKind.TriggerForIncompleteCompletions, But we could just start with CompletionTriggerKind.Invoked first.

I also look into making the JDT Completion Computer UI free code. But I think thats a huge under taking. We could do a better extension point or similar in jdt.ls if we think from scratch. Thats my person view. What do you all think ?

**Reply by rgrunber** [reply]
Chain completion is almost entirely in jdt.core.manipulation. The only part the remains in jdt.ui are the parts that initialize the java context for completion at an offset, and that build up the first set of completions. All of that can probably be done in JDT-LS and the rest of the logic is in jdt.core.manipulation. I think I had a PR for this on the JDT-LS side but can't seem to find it :\
